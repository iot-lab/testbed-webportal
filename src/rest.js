import axios from 'axios'

const API_URL = `https://${process.env.VUE_APP_IOTLAB_HOST}/api/`

export const iotlab = {

  api: axios.create({baseURL: API_URL, auth: JSON.parse(localStorage.getItem('apiAuth') || '{}')}),

  create (username, password) {
    this.api = axios.create({
      baseURL: API_URL,
      auth: {
        username: username,
        password: password,
      },
    })
  },

  // USERS API

  async signup (user, nbUsers) {
    if (nbUsers) {
      // create multiple user accounts
      await iotlab.api.post(`/users?nbusers=${nbUsers}`, user)
    } else {
      // create single user account
      await iotlab.api.post('/users', user)
    }
  },

  async activateAccount (hash) {
    await iotlab.api.post('/users/activate/', {'hash': hash})
  },

  async deleteUser (login = undefined) {
    if (login) {
      await iotlab.api.delete(`/users/${login}`)
    } else {
      await iotlab.api.delete(`/user`)
    }
  },

  async resetPassword (email) {
    await iotlab.api.post(`/users/reset_password`, {'email': email})
  },

  async changePassword (oldPwd, newPwd, confirmPwd) {
    await iotlab.api.put('/user/password', {
      'old_password': oldPwd,
      'new_password': newPwd,
      'confirm_new_password': confirmPwd,
    })
  },

  async getSSHkeys () {
    return iotlab.api.get('/user/keys').then(resp => resp.data.sshkeys)
  },

  async modifySSHkeys (keys) {
    return iotlab.api.post('/user/keys', {'sshkeys': keys})
  },

  async deleteSSHkey (id) {
    return iotlab.api.delete(`/user/keys/${id}`)
  },

  async getUsers ({status, group, search} = {}) {
    let params = {}
    if (status) {
      params.status = status
    }
    if (group) {
      params.group = group
    }
    if (search) {
      params.search = search
    }
    return iotlab.api.get('/users', {params: params}).then(resp => resp.data)
  },

  async getUserInfo (login = undefined) {
    if (login) {
      return iotlab.api.get(`/users/${login}`).then(resp => resp.data)
    } else {
      return iotlab.api.get('/user').then(resp => resp.data)
    }
  },

  async setUserInfo (user, login = undefined) {
    if (login) {
      return iotlab.api.put(`/users/${login}`, user)
    } else {
      return iotlab.api.put('/user', user)
    }
  },

  async activateUser (login) {
    return iotlab.api.put(`/users/${login}`, {'status': 'active'})
  },

  async deactivateUser (login) {
    return iotlab.api.put(`/users/${login}`, {'status': 'inactive'})
  },

  async setUserGroups (login, groups) {
    return iotlab.api.put(`/users/${login}`, {'groups': groups})
  },

  async sendMail (recipients, subject, message) {
    await iotlab.api.post('/users/email', {
      recipients: recipients,
      subject: subject,
      body: message,
    })
  },

  // EXPERIMENT API

  async getUserExperimentsCount () {
    return iotlab.api.get('/experiments/total').then(resp => resp.data)
  },

  async getAllExperimentsCount (user) {
    if (user) {
      return iotlab.api.get(`/experiments/total/users/${user}`).then(resp => resp.data)
    } else {
      return iotlab.api.get(`/experiments/total/all`).then(resp => resp.data)
    }
  },

  async getAllExperiments ({limit, offset, user, state = 'Terminated,Error,Running,Finishing,Resuming,toError,Waiting,Launching,Hold,toLaunch,toAckReservation,Suspended'} = {}) {
    let params = {}
    if (state) {
      params.state = state
    }
    if (limit) {
      params.limit = Math.min(limit, 500) // IoT-LAB api restrict queries to 500 entries
    }
    if (offset) {
      params.offset = offset
    }
    if (user === '@self') {
      return iotlab.api.get('/experiments', {params: params}).then(resp => resp.data.items)
    } else if (user) {
      return iotlab.api.get(`/experiments/users/${user}`, {params: params}).then(resp => resp.data.items)
    } else {
      return iotlab.api.get('/experiments/all', {params: params}).then(resp => resp.data.items)
    }
  },

  async getRunningExperiments () {
    return iotlab.api.get('/experiments/running').then(resp => resp.data.items)
  },

  async getExperiment (id) {
    return iotlab.api.get(`/experiments/${id}`).then(resp => resp.data)
  },

  async getExperimentArchive (id) {
    return iotlab.api.get(`/experiments/${id}/data`, {responseType: 'arraybuffer'}).then(resp => resp.data)
  },

  async getExperimentDeployment (id) {
    return iotlab.api.get(`/experiments/${id}/deployment`).then(resp => resp.data)
  },

  async stopExperiment (id) {
    return iotlab.api.delete(`/experiments/${id}`)
  },

  async submitExperiment ({type, name, duration, reservation, nodes, firmwareassociations, firmwares, profileassociations, profiles} = {}) {
    const formData = new FormData()
    formData.append('experiment.json', JSON.stringify({
      type: type,
      duration: duration,
      reservation: reservation,
      name: name,
      nodes: nodes,
      firmwareassociations: firmwareassociations,
      profileassociations: profileassociations,
    }))
    // attached firmware names should be unique
    let attachedFirmwares = []
    for (let asso of firmwareassociations || []) {
      if (asso.firmwarename in firmwares && !attachedFirmwares.includes(asso.firmwarename)) {
        formData.append(asso.firmwarename, new Blob([firmwares[asso.firmwarename]], {type: 'application/octet-stream'}), asso.firmwarename)
        attachedFirmwares.push(asso.firmwarename)
      }
    }
    return iotlab.api.post('/experiments', formData).then(resp => resp.data)
  },

  async checkFirmware (fileString) {
    const formData = new FormData()
    formData.append('firmware.bin', fileString)
    return iotlab.api.post('/firmwares/checker', formData).then(resp => resp.data)
  },

  async flashFirmware (id, nodes, fileString) {
    const formData = new FormData()
    formData.append(`experiment${id}.json`, JSON.stringify(nodes))
    formData.append('firmware.bin', new Blob([fileString], {type: 'application/octet-stream'}), 'firmware.bin')
    return iotlab.api.post(`/experiments/${id}/nodes/flash`, formData).then(resp => resp.data)
  },

  async flashResourcesFirmware (id, nodes, name) {
    return iotlab.api.post(`/experiments/${id}/nodes/flash/${name}`, nodes).then(resp => resp.data)
  },

  async updateMonitoring (id, nodes, name) {
    return iotlab.api.post(`/experiments/${id}/nodes/monitoring/${name}`, nodes).then(resp => resp.data)
  },

  async updateExperimentMobilityCircuit (id, nodes, name) {
    return iotlab.api.post(`/experiments/${id}/robots/mobility/${name}`, nodes).then(resp => resp.data)
  },

  async reloadExperiment (id) {
    return iotlab.api.post(`/experiments/${id}/reload`, {}).then(resp => resp.data)
  },

  // USER GROUPS

  async getUserGroups () {
    return iotlab.api.get('/groups').then(resp => resp.data)
  },

  async createGroup (name) {
    return iotlab.api.post(`/groups`, {name: name})
  },

  async deleteGroup (group) {
    return iotlab.api.delete(`/groups/${group}`)
  },

  // OTHER API

  async getSites () {
    return iotlab.api.get('/sites').then(resp => resp.data.items)
  },

  async getSitesDetails () {
    return iotlab.api.get('/sites/details').then(resp => resp.data.items)
  },

  async getNodes () {
    return iotlab.api.get('/nodes').then(resp => resp.data.items)
  },

  async updateNodesProperties () {
    return iotlab.api.put('/nodes')
  },

  async updateNodesPropertiesOldApi () {
    const OLD_API = `https://${process.env.VUE_APP_IOTLAB_HOST}/rest/`
    const oldApi = axios.create({baseURL: OLD_API, auth: JSON.parse(localStorage.getItem('apiAuth') || '{}')})
    return oldApi.post('/admin/nodes')
  },

  async getNodesIds () {
    return iotlab.api.get('/nodes/ids').then(resp => resp.data.items)
  },

  async getExperimentNodes (id) {
    return iotlab.api.get(`/experiments/${id}/nodes`).then(resp => resp.data.items)
  },

  async getExperimentToken (id) {
    return iotlab.api.get(`/experiments/${id}/token`).then(resp => resp.data.token)
  },

  async sendNodesCommand (id, cmd, nodes) {
    // cmd in [start, stop, reset, flash-idle, profile-reset, debug-start, debug-stop]
    return iotlab.api.post(`/experiments/${id}/nodes/${cmd}`, nodes).then(resp => resp.data)
  },

  // STORES (firmwares, profiles, etc)

  async getMobilityCircuits () {
    return iotlab.api.get('/mobilities/circuits').then(resp => resp.data.items)
  },

  async getMobilityCircuit (name) {
    return iotlab.api.get(`/mobilities/circuits/${name}`).then(resp => resp.data)
  },

  async getRobotsSiteMapImage (site) {
    return iotlab.api.get(`/robots/${site}/map/image`, {
      responseType: 'arraybuffer',
      headers: {
        'Access-Control-Allow-Origin': '*',
      },
    }).then(resp => Buffer.from(resp.data, 'binary').toString('base64')
    )
  },

  async getRobotsSiteMapConfig (site) {
    return iotlab.api.get(`/robots/${site}/map/config`).then(resp => resp.data)
  },

  async createMobilityCircuit (circuit) {
    return iotlab.api.post(`/mobilities/circuits`, circuit)
  },

  async updateMobilityCircuit (name, circuit) {
    return iotlab.api.put(`/mobilities/circuits/${name}`, circuit)
  },

  async deleteMobilityCircuit (name) {
    return iotlab.api.delete(`/mobilities/circuits/${name}`)
  },

  async getMonitoringProfiles () {
    return iotlab.api.get('/monitoring').then(resp => resp.data)
  },

  async getMonitoringProfile (name) {
    return iotlab.api.get(`/monitoring/${name}`).then(resp => resp.data)
  },

  async updateMonitoringProfile (name, profile) {
    return iotlab.api.put(`/monitoring/${name}`, profile)
  },

  async createMonitoringProfile (profile) {
    return iotlab.api.post(`/monitoring`, profile)
  },

  async deleteMonitoringProfile (name) {
    return iotlab.api.delete(`/monitoring/${name}`)
  },

  async getFirmwares () {
    return iotlab.api.get('/firmwares').then(resp => resp.data)
  },

  async getFirmware (name) {
    return iotlab.api.get(`/firmwares/${name}`).then(resp => resp.data)
  },

  async getFirmwareFile (name) {
    return iotlab.api.get(`/firmwares/${name}/file`, {responseType: 'arraybuffer'}).then(resp => resp.data)
  },

  async updateFirmware (name, firmware, fileString) {
    const formData = new FormData()
    formData.append(`firmware.json`, JSON.stringify(firmware))
    if (fileString) {
      formData.append('firmware.bin', new Blob([fileString], {type: 'application/octet-stream'}), 'firmware.bin')
    }
    return iotlab.api.put(`/firmwares/${name}`, formData).then(resp => resp.data)
  },

  async createFirmware (firmware, fileString) {
    const formData = new FormData()
    formData.append(`firmware.json`, JSON.stringify(firmware))
    formData.append('firmware.bin', new Blob([fileString], {type: 'application/octet-stream'}), 'firmware.bin')
    return iotlab.api.post(`/firmwares`, formData).then(resp => resp.data)
  },

  async deleteFirmware (name) {
    return iotlab.api.delete(`/firmwares/${name}`)
  },
}
