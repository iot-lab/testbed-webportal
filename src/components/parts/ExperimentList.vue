<template>
  <div>
    <h4 class="text-secondary" v-if="title && experiments.length">{{title}}</h4>
    <table class="table table-striped table-sm" v-if="experiments.length">
      <thead>
        <tr>
          <th>Id</th>
          <th v-if="user != '@self'">User</th>
          <th>Name</th>
          <th>Date</th>
          <th style="text-align: center">Duration</th>
          <th style="text-align: right">Nodes</th>
          <th style="text-align: right">State</th>
          <th width="10px"></th>
        </tr>
      </thead>
      <tbody>
        <tr v-for="exp in experiments">
          <td><router-link :to="{name: 'experimentDetails', params: { id: exp.id }}">{{exp.id}}</router-link></td>
          <td v-if="user != '@self'"><router-link :to="{name: 'users', query: { search: exp.user }}">{{exp.user}}</router-link></td>
          <td>{{exp.name}}</td>
          <td>{{startDate(exp) | formatDateTime}}</td>
          <td :class="{'durationProgress': showProgress(exp)}" :style="`text-align: center; --progress: ${experimentProgress(exp)}%`">
            {{(exp.effective_duration || exp.submitted_duration) | humanizeDuration}}
            <small class="text-dark" v-if="showProgress(exp)">({{experimentProgress(exp)}}%)</small>
          </td>
          <td style="text-align: right">{{exp.nb_nodes}}</td>
          <td style="text-align: right"><span class="badge badge-state" :class="exp.state | stateBadgeClass">{{exp.state}}</span></td>
          <td>
            <div class="dropdown">
              <button class="btn-link border-0 text-secondary dropdown-toggle" type="button" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"></button>
              <div class="dropdown-menu dropdown-menu-right">
                <router-link class="dropdown-item" :to="{name: 'experimentDetails', params: { id: exp.id }}">
                  <i class="fa fa-fw fa-eye"></i> View details
                </router-link>
                </a>
                <a class="dropdown-item text-danger" href="" @click.prevent="stopExperiment(exp)"
                  v-if="expStates.stoppable.includes(exp.state)">
                  <i class="fa fa-fw fa-ban"></i> Cancel
                </a>
                <a class="dropdown-item" href="" @click.prevent="startExperiment(exp)"
                  v-if="exp.state === 'Waiting'">
                  <i class="fa fa-fw fa-play"></i> Start
                </a>
                <a class="dropdown-item" href="" @click.prevent="reloadExperiment(exp)"
                  v-if="expStates.completed.includes(exp.state)">
                  <i class="fa fa-fw fa-refresh"></i> Restart
                </a>
                <!-- <a class="dropdown-item" href="" @click.prevent="this.alert('todo')">
                  <i class="fa fa-fw fa-clone"></i> Clone
                </a> -->
              </div>
            </div>

          </td>
        </tr>
        <tr v-show="!spinner && experiments.length === 0">
          <td colspan="7" class="font-italic">No items to display</td>
        </tr>
      </tbody>
      <tfoot v-if="total > experiments.length && !spinner">
        <tr>
          <td colspan="7" class="font-italic cursor btn-link" @click="loadMore(step)">Load more...</td>
        </tr>
      </tfoot>
    </table>
    <p v-show="spinner">
      <i class="fa fa-spinner fa-spin fa-fw"></i>
    </p>
  </div>
</template>

<script>
import { iotlab } from '@/rest'
import { experimentStates } from '@/assets/js/iotlab-utils'
import { sleep } from '@/utils'

export default {
  name: 'ExperimentList',

  props: {
    title: {
      // Optional title
      type: String,
      default: '',
    },
    user: {
      // Filter experiments by user. Leave empty for all users. Or '@self' for current (logged in) user.
      type: String,
    },
    state: {
      // Comma separated list of states to be included in the query
      // or use meta state:
      // 'all_scheduled' -> 'Running,Finishing,Resuming,toError,Waiting,Launching,Hold,toLaunch,toAckReservation,Suspended'
      // 'all_terminated' -> 'Terminated,Stopped,Error'
      type: String,
    },
    total: {
      // Total number of experiments
      // (mandatory to display experiments with bigger ID first, since the API results are ordered by ID ascending)
      type: Number,
      default: 0,
    },
    show: {
      // Max number of items to load intially
      type: Number,
      default: 500,
    },
    step: {
      // Number of items to load on "load more" action
      type: Number,
      default: 50,
    },
    started: {
      // Started experiment counter, increment to force component to refresh
      type: Number,
      default: 0,
    },
  },

  data () {
    return {
      spinner: false,
      states: undefined,
      expStates: experimentStates,
      experiments: [],
      polling: false,
    }
  },

  async created () {
    this.states = this.state
    if (this.states === 'all_scheduled') {
      this.states = 'Running,Finishing,Resuming,toError,Waiting,Launching,Hold,toLaunch,toAckReservation,Suspended'
    }
    if (this.states === 'all_terminated') {
      this.states = 'Terminated,Stopped,Error'
    }

    this.loadMore(this.show)
  },

  destroyed () {
    this.stopPolling()
  },

  watch: {
    total: function (newTotal, oldTotal) {
      // experiment total changed, let's refresh including new xp
      this.refresh(newTotal - oldTotal)
    },
    started: function () {
      // started counter has been touched, let's refresh
      this.refresh()
    },
  },

  methods: {

    showProgress: exp => exp.state === 'Running',

    experimentProgress (exp) {
      // Gives the percentage of progress
      if (experimentStates.completed.includes(exp.state)) return 0
      return Math.min(100, Math.round(100 * exp.effective_duration / exp.submitted_duration))
    },

    startDate (exp) {
      return [exp.start_date, exp.scheduled_date, exp.submission_date].find(date => date !== '1970-01-01T00:00:00Z')
    },

    async loadMore (qty) {
      this.spinner = true

      this.experiments = this.experiments.concat((await iotlab.getAllExperiments({
        user: this.user,
        state: this.states,
        offset: Math.max(this.total - this.experiments.length - qty, 0),
        limit: Math.min(this.total - this.experiments.length, qty),
      })).sort((exp1, exp2) => exp2.id - exp1.id)) // order by reverse ID

      this.spinner = false

      if (this.experiments.some(exp => experimentStates.scheduled.includes(exp.state))) {
        this.startPolling()
      }
    },

    async refresh (more = 0) {
      let previous = this.experiments
      this.experiments = (await iotlab.getAllExperiments({
        user: this.user,
        state: this.states,
        // offset: Math.max(this.total - this.experiments.length - qty, 0),
        offset: Math.max(this.total - this.experiments.length - more, 0),
        limit: Math.max(this.experiments.length, this.show) + more,
      })).sort((exp1, exp2) => exp2.id - exp1.id) // order by reverse ID

      if (this.experiments.some(exp => experimentStates.scheduled.includes(exp.state))) {
        this.startPolling()
      } else {
        this.stopPolling()
      }

      if (this.experiments.length < previous.length) {
        // some experiments have ended, let's notify
        this.$emit('completed')
      }
    },

    startPolling () {
      // create watchdog to poll for scheduled experiments update
      if (!this.polling) this.polling = setInterval(() => this.refresh(), 5000)
    },

    stopPolling () {
      clearInterval(this.polling)
      this.polling = false
    },

    async stopExperiment (exp, confirmed = false) {
      if (!confirmed && !confirm('Cancel this experiment?')) return
      try {
        await iotlab.stopExperiment(exp.id)
        this.$notify({text: `Experiment ${exp.id} stopping`, type: 'success'})
        if (exp.state === 'Running') exp.state = 'Finishing'
      } catch (err) {
        this.$notify({text: err.message, type: 'error'})
      }
    },

    async startExperiment (exp) {
      if (!confirm('Resubmit now and cancel this experiment?')) return
      await this.stopExperiment(exp, true)
      await this.reloadExperiment(exp)
    },

    async reloadExperiment (exp) {
      try {
        let newExp = await iotlab.reloadExperiment(exp.id)
        this.$notify({text: `Experiment ${newExp.id} submitted`, type: 'success'})
        await sleep(200)
        this.$emit('started')
      } catch (err) {
        this.$notify({text: err.message, type: 'error'})
      }
    },
  },
}
</script>

<style scoped>
.dropdown .dropdown-toggle:hover {
  filter: brightness(0.5);
}

.durationProgress {
  position: relative;
  z-index: 1;
  text-align: center;
}
.durationProgress::after {
  content: '';
  position: absolute;
  top: 5px;
  left: 0;
  height: calc(100% - 10px);
  width: var(--progress);
  background-color: #17a2b890;
  border-left: 0.5px solid #17a2b850;
  border-top: 0.5px solid #17a2b850;
  border-bottom: 0.5px solid #17a2b850;
  border-radius: 3px 0 0 3px;
  z-index: -1;
}
.durationProgress::before {
  content: '';
  position: absolute;
  top: 5px;
  left: 0;
  height: calc(100% - 10px);
  width: 100%;
  background-color: var(--light);
  border: 0.5px solid lightgrey;
  border-radius: 3px;
  z-index: -1;
}
</style>
