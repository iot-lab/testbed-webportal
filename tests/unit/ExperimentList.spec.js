import { shallowMount, createLocalVue } from '@vue/test-utils'
import VueRouter from 'vue-router'
import ExperimentList from '@/components/ExperimentList'
import { iotlab } from '@/rest'
jest.mock('@/rest')

// [{"submitted_duration":60,"submission_date":"2018-07-25T15:37:26Z","nb_nodes":15,"effective_duration":0,"name":"","stop_date":"1970-01-01T00:00:00Z","id":13166,"state":"Waiting","scheduled_date":"1970-01-01T00:00:00Z","user":"bformet","start_date":"1970-01-01T00:00:00Z"},{"submitted_duration":60,"submission_date":"2018-08-29T23:14:27Z","nb_nodes":10,"effective_duration":1,"name":"","stop_date":"1970-01-01T00:00:00Z","id":13226,"state":"Running","scheduled_date":"2018-08-29T23:14:29Z","user":"bformet","start_date":"2018-08-29T23:14:29Z"}]

const localVue = createLocalVue()
localVue.use(VueRouter)
const router = new VueRouter()

describe('ExperimentList.vue', () => {

  it('should match snapshot', () => {
    const wrapper = shallowMount(ExperimentList, {localVue, router})
    expect(wrapper.element).toBeDefined()
    expect(wrapper.element).toMatchSnapshot()
  })

})
