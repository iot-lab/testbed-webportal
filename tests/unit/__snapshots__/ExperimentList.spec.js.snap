// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExperimentList.vue should match snapshot 1`] = `
<div>
  <!---->
   
  <!---->
   
  <div
    style="display: none;"
  >
    <i
      class="fa fa-spinner fa-spin fa-fw"
    />
     
    <i>
      loading experiments
    </i>
  </div>
</div>
`;
